File: leetcode/0205.py
class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:
        mapping_s_to_t = {}
        mapping_t_to_s = {}
        for a, b in zip(s, t):
            if a not in mapping_s_to_t:
                if b not in mapping_t_to_s:
                    mapping_s_to_t[a] = b
                    mapping_t_to_s[b] = a
                else:
                    return False
            else:
                if mapping_s_to_t[a] != b:
                    return False
        return True

File: leetcode/0205.cpp
class Solution {
public:
    bool isIsomorphic(string s, string t) {
        unordered_map<char, char> mapping_s_to_t, mapping_t_to_s;
        bool find_t, find_s;
        for(auto i = 0; i < s.size(); ++i)
        {
            find_s = mapping_s_to_t.find(s[i]) != mapping_s_to_t.end();
            find_t = mapping_t_to_s.find(t[i]) != mapping_t_to_s.end();
            if (find_s || find_t)
            {
                if ((mapping_s_to_t[s[i]] != t[i]) || (mapping_t_to_s[t[i]] != s[i]))
                    return false;
            }
            else
            {
                mapping_s_to_t[s[i]] = t[i];
                mapping_t_to_s[t[i]] = s[i];
            }
        }
        return true; 
    }
};
